### PR TITLE
[Enhance] Enhance vis-pipeline tool

### PR DESCRIPTION
**Pull request by Ezra-Yu**
## Motivation

Enhance vis-pipeline tool refer to #561

## Modification

1. add 'transformed' mode to vis transformed images.
2. modify 'pipeline' mode to vis intermediate images.
3. replace example pictures with url link, remove real pictures docs/_static/xxxx

## BC-breaking (Optional)

the behave of mode 'pipeline'  has changed.

## Use cases (Optional)

1. 'transformed' mode 
 ```shell
  python ./tools/visualizations/vis_pipeline.py ./configs/resnet/resnet50_8xb32_in1k.py --show --mode transformed
  ```
  <div align=center></div>

2. pipeline' mode
  ```shell
  python ./tools/visualizations/vis_pipeline.py configs/swin_transformer/swin_base_224_b16x64_300e_imagenet.py --phase train --adaptive --mode pipeline --show
  ```

  <div align=center></div>

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by yingfhu — CHANGES_REQUESTED** [review]

File: tools/visualizations/vis_pipeline.py [hunk 1]
```diff
@@ -143,43 +148,80 @@ def build_dataset_pipeline(cfg, phase):
...
+
+    return dataset, pipelines
+
+
+def concat_imgs(args, imgs: List[np.ndarray], steps: List[str]) -> np.ndarray:
+    """Concat list of pictures into a single big picture, align height here."""
+    ori_shapes = [img.shape for img in imgs]
+    # if src image is too large or too small, it will make the text
+    # information difficult to confirm, so here add adaptive adjustment.
+    if args.adaptive:
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:164]: Can add a warning about suggestion of using `--adaptive` for small images.

**Review by yingfhu — APPROVED** [review]
LGTM

File: tools/visualizations/vis_pipeline.py [hunk 2]
```diff
@@ -143,43 +148,80 @@ def build_dataset_pipeline(cfg, phase):
...
-
-    put_img(board, left_img, (int(board_w // 2), int(board_h // 2)))
-    put_img(board, right_img,
-            (int(board_w // 2) + board_w + GAP // 2, int(board_h // 2)))
-    return board
-
-
-def adaptive_size(mode, image, min_edge_length, max_edge_length):
+def adaptive_size(image, min_edge_length, max_edge_length, src_shape=None):
     """rescale image if image is too small to put text like cifra."""
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:220]: cifar

File: tools/visualizations/vis_pipeline.py (hunk 1 above)
Comment (Ezra-Yu) [reply tools/visualizations/vis_pipeline.py:164]: Do you mean when reading some pictures, which have very small or very lager resolution, then we give a warning like
"the src picture have size of xxx, it is better to add `--adaptive` to adaptively resize to show?"

Comment (Ezra-Yu) [reply tools/visualizations/vis_pipeline.py:169]: done.

File: tools/visualizations/vis_pipeline.py (hunk 2 above)
Comment (Ezra-Yu) [reply tools/visualizations/vis_pipeline.py]: done,

File: tools/visualizations/vis_pipeline.py [hunk 3]
```diff
@@ -143,43 +154,103 @@ def build_dataset_pipeline(cfg, phase):
...
+        for pipeline_cfg in origin_pipeline
+    }
+
+    return dataset, pipelines
+
+
+def prepare_imgs(args, imgs: List[np.ndarray], steps=None):
+    """prepare the showing picture."""
+    ori_shapes = [img.shape for img in imgs]
+    # adaptivly adjustment to rescale pictures
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:168]: adaptively

File: tools/visualizations/vis_pipeline.py [hunk 4]
```diff
@@ -143,43 +154,103 @@ def build_dataset_pipeline(cfg, phase):
...
+        imgs[i] = cv2.putText(
+            img=img,
+            text=str(ori_shapes[i]),
+            org=(pic_horizontal_gap, max_height + int(text_height * 1.5)),
+            fontFace=cv2.FONT_HERSHEY_TRIPLEX,
+            fontScale=font_size,
+            color=(255, 0, 0),
+            lineType=1)
+
+    # High alignment for concatenating
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:243]: height

File: tools/visualizations/vis_pipeline.py [hunk 5]
```diff
@@ -190,49 +261,58 @@ def adaptive_size(mode, image, min_edge_length, max_edge_length):
...
-
-    if mode == 'concat':
-        image = concat(src_image, trans_image)
-    elif mode == 'original':
-        image = src_image
-    elif mode == 'pipeline':
-        image = trans_image
+    pipeline_images = [src_image]
+
+    # get intermediates images through pipelines
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:272]: intermediate

File: tools/visualizations/vis_pipeline.py [hunk 6]
```diff
@@ -190,49 +261,58 @@ def adaptive_size(mode, image, min_edge_length, max_edge_length):
...
+    if args.mode in ['transformed', 'concat', 'pipeline']:
+        for pipeline in pipelines.values():
+            item = pipeline(item)
+            trans_image = copy.deepcopy(item['img'])
+            trans_image = np.ascontiguousarray(trans_image, dtype=np.uint8)
+            pipeline_images.append(trans_image)
+
+    # concat all the images to show, depandding on 'mode'
```
Comment (yingfhu) [inline tools/visualizations/vis_pipeline.py:280]: concatenate images to be showed according to `mode`

File: docs/en/tools/visualization.md [hunk 7]
```diff
@@ -2,29 +2,29 @@
...
+    [--show] \
+    [--adaptive] \
+    [--min-edge-length ${MIN_EDGE_LENGTH}] \
+    [--max-edge-length ${MAX_EDGE_LENGTH}] \
+    [--bgr2rgb] \
+    [--window-size ${WINDOW_SIZE}] \
+    [--cfg-options ${CFG_OPTIONS}]
 ```
 
 **Description of all arguments**：
```
Comment (yingfhu) [inline docs/en/tools/visualization.md:30]: #35 to be visualized
#39 If set, adaptively resize images for better visualization.

File: docs/en/tools/visualization.md [hunk 8]
```diff
@@ -41,39 +41,48 @@ python tools/visualizations/vis_pipeline.py \
 - `--max-edge-length`: The maximum edge length, used when `--adaptive` is set. When any side of the picture is larger than `${MAX_EDGE_LENGTH}`, the picture will be reduced while keeping the aspect ratio unchanged, and the long side will be aligned to `${MAX_EDGE_LENGTH}`. If not specified, it will be set to 1000.
 - `--bgr2rgb`: If set, flip the color channel order of images.
 - `--window-size`: The shape of the display window. If not specified, it will be set to `12*7`. If used, it must be in the format `'W*H'`.
+- `--cfg-options` : Modifications to the configuration file, refer to [Tutorial 1: Learn about Configs](https://mmclassification.readthedocs.io/en/latest/tutorials/config.html).
 
 ```{note}
```
Comment (yingfhu) [inline docs/en/tools/visualization.md:46]: It is hard to read notes in codeblock.

File: docs/en/tools/visualization.md [hunk 9]
```diff
@@ -41,39 +41,48 @@ python tools/visualizations/vis_pipeline.py \
...
-   ```
+  ```shell
+  python ./tools/visualizations/vis_pipeline.py ./configs/resnet/resnet50_8xb32_in1k.py --show --mode transformed
+  ```
 
-   <div align=center><img src="../_static/image/tools/visualization/pipeline-concat.jpg" style=" width: auto; height: 40%; "></div>
+  <div align=center><img src="https://user-images.githubusercontent.com/18586273/146117553-8006a4ba-e2fa-4f53-99bc-42a4b06e413f.jpg" style=" width: auto; height: 40%; "></div>
 
-3. Visualize 100 original pictures in the `CIFAR100` validation set, then display and save them in the `./tmp` folder：
+3. In **'concat'** mode, visualize 10 comparison pictures in the `ImageNet` train set and save them in the `./tmp` folder：
```
Comment (yingfhu) [inline docs/en/tools/visualization.md:71]: visualize 10 pairs of origin and transformed images for comparison.

File: docs/en/tools/visualization.md (hunk 7 above)
Comment (Ezra-Yu) [reply docs/en/tools/visualization.md:30]: done.

**Pull request merged**